### PR TITLE
#2633 -  Child care cost calculation PT

### PR DESCRIPTION
**Pull request by dheepak-aot**
## Child Care Cost calculation PT

- [x] Calculation of eligible dependents happens irrespective of the `student relationship status` and hence this calculation is moved before the exclusive gateway and removed repetition of the code in 2 places.
image

- [x] **Age calculation with study start date:** Age of the student dependents are now updated to be calculated based on the `offering study start date` and NOT the current date.
image

- [x] Added workflow logic to get the count of dependents on each category, `Age 11 Years or under` and `Age 12 Years or over and declared on taxes for disability` and total number of dependents which fall under either of these categories `Eligible dependents for child care cost`

image

- [x] Workflow specific calculated fields created with respect to `Child care cost for 11 years and under` and `Child care cost for 12 years and over and `  based on the student input and the number of dependents they have in each category.

These numbers will lead us to child care cost(without considering maximum limits which will be applied later)

image

- [x] Calculating the final child care cost value `calculatedDataTotalChildCareCost` based on offering weeks (and also course load)

image

- [x] Added E2E Tests for child care cost calculation under a new folder `costs`.

- [x] Hiding child care cost questions for student with NO dependents

image

### Formatter difference

There is a formatter difference observed in one place in form.io. It is not part of the PR.

image

### CODE REVIEW

File: sources/packages/forms/src/form-definitions/sfaa2021-22.json [hunk 1]
```diff
@@ -1586,7 +1587,7 @@
                 "eq": "",
                 "json": ""
               },
-              "customConditional": "show =\r\n  (!!data.selectedLocation &&\r\n  !!!data.mySchoolIsNotListed) ||\r\n  !!data.isReadOnly;",
+              "customConditional": "show =\r\n  (!!data.selectedLocation &&\r\n  !!!data.mySchoolIsNotListed) || !!data.isReadOnly;",
```
Comment (dheepak-aot) [inline sources/packages/forms/src/form-definitions/sfaa2021-22.json:1590]: These changes are not part of the PR. It is due to difference in formatter.

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn [hunk 2]
```diff
@@ -579,39 +391,276 @@
...
+        </zeebe:ioMapping>
+      </bpmn:extensionElements>
+      <bpmn:incoming>Flow_0dqxcli</bpmn:incoming>
+      <bpmn:outgoing>Flow_0kv24s2</bpmn:outgoing>
+    </bpmn:intermediateThrowEvent>
+    <bpmn:intermediateThrowEvent id="Event_1i1k5zk" name="Calculate Eligible Depenents for Child Care Cost">
+      <bpmn:extensionElements>
+        <zeebe:ioMapping>
+          <zeebe:output source="=if studentDataDependants = null then 0 else&#10;  count(studentDataDependants[&#10;      // Dependant who is between 0 and 11 years until they become 12 years old.&#10;      // i.e. until the day(study start date) before their 12th birthday the dependant &#10;      // is considered to be in this category.&#10;      date(item.dateOfBirth) &#62; (date(offeringStudyStartDate) - duration(&#34;P12Y&#34;))]&#10;  )" target="calculatedDataDependants11YearsOrUnder" />
```
Comment (andrepestana-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn:656]: I think we should rephrase the sentence:
`// Dependant who is between 0 and 11 years until they become 12 years old.`
to:
`// Dependant who is between 0 and 11 years old until the offering start date.`

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/costs/parttime-assessment-2021-2022-costs-child-care-cost.e2e-spec.ts [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  it(
+    "Should calculate child care costs as 0 when student has one or more dependents on both categories(11 years or under and over 12 years " +
+      " or over and declared on taxes for disability) but child care costs not provided in student application.",
+    async () => {
+      // Arrange
+      const assessmentConsolidatedData =
+        createFakeConsolidatedPartTimeData(PROGRAM_YEAR);
+      assessmentConsolidatedData.offeringWeeks = 8;
+      assessmentConsolidatedData.offeringCourseLoad = 20;
+      // Creates 4 eligible and 4 not eligible dependents.
```
Comment (andrepestana-aot) [inline sources/packages/backend/workflow/test/2021-2022/parttime-assessment/costs/parttime-assessment-2021-2022-costs-child-care-cost.e2e-spec.ts:140]: Shouldn't it be "1 eligible and 1 not eligible" ?

File: sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts [hunk 4]
```diff
@@ -93,3 +104,44 @@ export function createFakeStudentDependentNotEligible(
...
+  switch (eligibility) {
+    case DependentChildCareEligibility.Eligible0To11YearsOld:
+      return {
+        dateOfBirth: addToDateOnlyString(studyStartDate, -11, "years"),
+        attendingPostSecondarySchool: YesNoOptions.No,
+        declaredOnTaxes: YesNoOptions.No,
+      };
+    case DependentChildCareEligibility.Eligible12YearsAndOver:
+      return {
+        dateOfBirth: addToDateOnlyString(studyStartDate, -12.1, "years"),
```
Comment (andrepestana-aot) [inline sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts:127]: Shouldn't `-12` be enough? As I understood it is 12 y.o. or over.

File: sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts [hunk 5]
```diff
@@ -93,3 +104,44 @@ export function createFakeStudentDependentNotEligible(
...
+/**
+ * Create a student dependant who is not eligible for child care costs.
+ * @param studyStartDate study start date of the offering.
+ * @returns dependant who is not eligible.
+ */
+export function createFakeStudentDependentNotEligibleForChildcareCost(
+  studyStartDate: Date | string,
+): StudentDependent {
+  return {
+    dateOfBirth: addToDateOnlyString(studyStartDate, -12.1, "years"),
```
Comment (andrepestana-aot) [inline sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts:143]: Same as https://github.com/bcgov/SIMS/pull/2816/files#r1486649054

**Review by sh16011993 — COMMENTED** [review]
I have a question from the story itself:

image

Here, the first two lines say which of the dependents are eligible. It says 12+ with disability are eligible.
 
And below the expectation removes disability and includes tax:

image

Can you please clarify on this @dheepak-aot. Maybe I am missing something. Thanks!

**Review by andrepestana-aot — COMMENTED** [review]
Good job! Please take a look at my comments.

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn:656]: @andrepestana-aot until they become 12 years old they are are considered as 11 years is what I was trying to convey here. 

What about `// Dependant who is between 0 and 11 years until they become 12 years old as on offering study start date.` ?

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/costs/parttime-assessment-2021-2022-costs-child-care-cost.e2e-spec.ts (hunk 3 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/test/2021-2022/parttime-assessment/costs/parttime-assessment-2021-2022-costs-child-care-cost.e2e-spec.ts:140]: It should be `// Creates 2 eligible dependents.` I will change it.

File: sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts (hunk 4 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/test/test-utils/factories/student-dependent.ts:127]: You are right, 12 is sufficient to create the dependent in that category. I kept it as 12 above just because it also qualifies.

I can change that.

File: sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+import {
+  DependentChildCareEligibility,
+  createFakeStudentDependentEligibleForChildcareCost,
+  createFakeStudentDependentNotEligibleForChildcareCost,
+} from "../../../test-utils/factories";
+
+describe(`E2E Test Workflow part-time-assessment-${PROGRAM_YEAR}-costs-child-care-costs.`, () => {
+  it(
+    "Should calculate total child care cost as sum of values in student application when student has one or more dependents " +
+      " on both categories (11 years or under and over 12 years or over and declared on taxes for disability) and " +
```
Comment (sh16011993) [inline sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts:15]: Very Minor: `over 12 years and declared on taxes`

File: sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+      // Child care cost expected to be the total sum of values entered in student application.
+      expect(calculatedAssessment.variables.calculatedDataChildCareCost).toBe(
+        totalChildcareCostEnteredInApplication,
+      );
+      // As the total sum of child care costs are beyond the maximum allowable limit
+      // the maximum limit must be considered towards assessed need.
+      expect(
+        calculatedAssessment.variables.calculatedDataTotalChildCareCost,
+      ).toBe(1822.4);
+      expect(
```
Comment (sh16011993) [inline sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts:123]: Isn't this expect redundant? If not, then we can add a similar expect in the first test above.

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/costs/parttime-assessment-2021-2022-costs-child-care-cost.e2e-spec.ts [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+      // Child care cost expected to be the total sum of values entered in student application.
+      expect(calculatedAssessment.variables.calculatedDataChildCareCost).toBe(
+        totalChildcareCostEnteredInApplication,
+      );
+      // As the total sum of child care costs are beyond the maximum allowable limit
+      // the maximum limit must be considered towards assessed need.
+      expect(
+        calculatedAssessment.variables.calculatedDataTotalChildCareCost,
+      ).toBe(1822.4);
+      expect(
```
Comment (sh16011993) [inline sources/packages/backend/workflow/test/2021-2022/parttime-assessment/costs/parttime-assessment-2021-2022-costs-child-care-cost.e2e-spec.ts:123]: https://github.com/bcgov/SIMS/pull/2816/files#r1486741841

File: sources/packages/backend/workflow/test/2023-2024/parttime-assessment/costs/parttime-assessment-2023-2024-costs-child-care-cost.e2e-spec.ts [hunk 9]
```diff
@@ -0,0 +1,211 @@
...
+      // Child care cost expected to be the total sum of values entered in student application.
+      expect(calculatedAssessment.variables.calculatedDataChildCareCost).toBe(
+        totalChildcareCostEnteredInApplication,
+      );
+      // As the total sum of child care costs are beyond the maximum allowable limit
+      // the maximum limit must be considered towards assessed need.
+      expect(
+        calculatedAssessment.variables.calculatedDataTotalChildCareCost,
+      ).toBe(1822.4);
+      expect(
```
Comment (sh16011993) [inline sources/packages/backend/workflow/test/2023-2024/parttime-assessment/costs/parttime-assessment-2023-2024-costs-child-care-cost.e2e-spec.ts:123]: https://github.com/bcgov/SIMS/pull/2816/files#r1486741841

File: sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+  it(
+    "Should calculate child care costs as 0 when student has one or more dependents on both categories(11 years or under and over 12 years " +
+      " or over and declared on taxes for disability) but child care costs not provided in student application.",
+    async () => {
+      // Arrange
+      const assessmentConsolidatedData =
+        createFakeConsolidatedPartTimeData(PROGRAM_YEAR);
+      assessmentConsolidatedData.offeringWeeks = 8;
+      assessmentConsolidatedData.offeringCourseLoad = 20;
+      // Creates 4 eligible and 4 not eligible dependents.
```
Comment (sh16011993) [inline sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts:140]: Very Minor: 2 eligible dependants.

File: sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+        calculatedAssessment.variables.calculatedDataTotalChildCareCost,
+      ).toBe(1822.4);
+      expect(
+        calculatedAssessment.variables.calculatedDataTotalChildCareCost,
+      ).toBeLessThan(
+        calculatedAssessment.variables.calculatedDataChildCareCost,
+      );
+    },
+  );
+
```
Comment (sh16011993) [inline sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts:130]: Just a suggestion: Since there are 3 branches below,

image

Maybe we can add another quick test to check offering weeks greater than 34 too!

Comment (dheepak-aot) [reply sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts:130]: I am not trying to cover every scenario in E2E test.

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn [hunk 12]
```diff
@@ -579,39 +391,276 @@
...
+          <zeebe:output source="=if studentDataDependants = null then 0 else&#10;  count(&#10;studentDataDependants[&#10;// 0-18 years of age: eligible.&#10;date(item.dateOfBirth) &#62;= (date(offeringStudyStartDate) - duration(&#34;P18Y&#34;))&#10;// 19-22 years of age: post-sec must be Yes or Tax is Yes.&#10;or (&#10;  date(item.dateOfBirth) &#60; (date(offeringStudyStartDate) - duration(&#34;P18Y&#34;))&#10;  and date(item.dateOfBirth) &#62;= (date(offeringStudyStartDate) - duration(&#34;P22Y&#34;))&#10;  and (item.attendingPostSecondarySchool = &#34;yes&#34; or item.declaredOnTaxes = &#34;yes&#34;)&#10;)&#10;// &#62; 22 years: eligible if Tax is Yes.&#10;or (&#10;  date(item.dateOfBirth) &#60; (date(offeringStudyStartDate) - duration(&#34;P22Y&#34;))&#10;  and item.declaredOnTaxes = &#34;yes&#34;&#10;)&#10;]&#10;  )" target="calculatedDataTotalEligibleDependants" />
+        </zeebe:ioMapping>
+      </bpmn:extensionElements>
+      <bpmn:incoming>Flow_0dqxcli</bpmn:incoming>
+      <bpmn:outgoing>Flow_0kv24s2</bpmn:outgoing>
+    </bpmn:intermediateThrowEvent>
+    <bpmn:intermediateThrowEvent id="Event_1i1k5zk" name="Calculate Eligible Depenents for Child Care Cost">
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn:653]: Typo in `Depenents`.

**Review by sh16011993 — COMMENTED** [review]
Nice Work @dheepak-aot . 👍 Very good test case coverage. Left some comments.

File: sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts (hunk 7 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts:123]: The second expect is being explicit about **considered child care costs** being less than total childcare costs entered in student application application(when exceeding maximum limit) .

The first test does not have child care cost exceeding maximum limit and hence it is not applicable to it.

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn [hunk 13]
```diff
@@ -579,39 +391,276 @@
...
+      <bpmn:incoming>Flow_1idr6ms</bpmn:incoming>
+      <bpmn:outgoing>Flow_0dqxcli</bpmn:outgoing>
+    </bpmn:intermediateThrowEvent>
+    <bpmn:intermediateThrowEvent id="Event_18l6de5" name="Calculate Eligible Dependents">
+      <bpmn:extensionElements>
+        <zeebe:ioMapping>
+          <zeebe:output source="=if studentDataDependants = null then 0 else&#10;  count(&#10;studentDataDependants[&#10;// 0-18 years of age: eligible.&#10;date(item.dateOfBirth) &#62;= (date(offeringStudyStartDate) - duration(&#34;P18Y&#34;))&#10;// 19-22 years of age: post-sec must be Yes or Tax is Yes.&#10;or (&#10;  date(item.dateOfBirth) &#60; (date(offeringStudyStartDate) - duration(&#34;P18Y&#34;))&#10;  and date(item.dateOfBirth) &#62;= (date(offeringStudyStartDate) - duration(&#34;P22Y&#34;))&#10;  and (item.attendingPostSecondarySchool = &#34;yes&#34; or item.declaredOnTaxes = &#34;yes&#34;)&#10;)&#10;// &#62; 22 years: eligible if Tax is Yes.&#10;or (&#10;  date(item.dateOfBirth) &#60; (date(offeringStudyStartDate) - duration(&#34;P22Y&#34;))&#10;  and item.declaredOnTaxes = &#34;yes&#34;&#10;)&#10;]&#10;  )" target="calculatedDataTotalEligibleDependants" />
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn:647]: I do recognize that we do not have a format for the Feel Expression but, as a suggestion, I would like to preserve some level of indentation as below.
```
if studentDataDependants = null then 0
else
  count(
    studentDataDependants[
      // 0-18 years of age: eligible.
      date(item.dateOfBirth) >= (date(offeringStudyStartDate) - duration("P18Y"))
      // 19-22 years of age: post-sec must be Yes or Tax is Yes.
      or (
        date(item.dateOfBirth) < (date(offeringStudyStartDate) - duration("P18Y"))
        and date(item.dateOfBirth) >= (date(offeringStudyStartDate) - duration("P22Y"))
        and (item.attendingPostSecondarySchool = "yes" or item.declaredOnTaxes = "yes")
      )
      // > 22 years: eligible if Tax is Yes.
      or (
        date(item.dateOfBirth) < (date(offeringStudyStartDate) - duration("P22Y"))
        and item.declaredOnTaxes = "yes")
    ]
  )
```

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/costs/parttime-assessment-2021-2022-costs-child-care-cost.e2e-spec.ts (hunk 8 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/test/2021-2022/parttime-assessment/costs/parttime-assessment-2021-2022-costs-child-care-cost.e2e-spec.ts:123]: https://github.com/bcgov/SIMS/pull/2816#discussion_r1486789749

File: sources/packages/backend/workflow/test/2023-2024/parttime-assessment/costs/parttime-assessment-2023-2024-costs-child-care-cost.e2e-spec.ts (hunk 9 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/test/2023-2024/parttime-assessment/costs/parttime-assessment-2023-2024-costs-child-care-cost.e2e-spec.ts:123]: https://github.com/bcgov/SIMS/pull/2816#discussion_r1486789749

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn [hunk 14]
```diff
@@ -95,84 +93,212 @@
...
+      <bpmn:incoming>Flow_173rbte</bpmn:incoming>
+      <bpmn:incoming>Flow_04xt4on</bpmn:incoming>
+      <bpmn:incoming>Flow_0wfngj4</bpmn:incoming>
+      <bpmn:outgoing>Flow_19cfg17</bpmn:outgoing>
+    </bpmn:exclusiveGateway>
+    <bpmn:intermediateThrowEvent id="Event_0re3zsg" name="Calculate Child Care Costs">
       <bpmn:extensionElements>
         <zeebe:ioMapping>
-          <zeebe:output source="=calculatedDataTotalEligibleDependants + 1" target="calculatedDataFamilySize" />
+          <zeebe:output source="=if (calculatedDataDependants11YearsOrUnder &#62; 0)&#10;  then studentDataDaycareCosts11YearsOrUnder&#10;else 0" target="calculatedDataDaycareCosts11YearsOrUnder" />
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn:251]: As a suggestion, the same logic can be simplified using a MAX. Same for the similar ones.

**Comment by dheepak-aot** [comment]
> I have a question from the story itself:
> 
> image
> Here, the first two lines say which of the dependents are eligible. It says 12+ with disability are eligible.
> 
> And below the expectation removes disability and includes tax:
> 
> image
> Can you please clarify on this @dheepak-aot. Maybe I am missing something. Thanks!

@sh16011993 as discussed, the `declaredOnTaxes` here means declared on taxes for disability and hence it does not remove the condition for disability

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn [hunk 15]
```diff
@@ -95,84 +93,212 @@
...
+    <bpmn:sequenceFlow id="Flow_0wfngj4" sourceRef="Event_1dvh1rf" targetRef="Gateway_1blaf95" />
+    <bpmn:sequenceFlow id="Flow_04xt4on" sourceRef="Event_1e56skq" targetRef="Gateway_1blaf95" />
+    <bpmn:sequenceFlow id="Flow_173rbte" sourceRef="Event_0xpdeqq" targetRef="Gateway_1blaf95" />
+    <bpmn:sequenceFlow id="Flow_0mlhznt" name="Offering weeks less than or equals 17" sourceRef="Gateway_1gcbtrb" targetRef="Event_1dvh1rf">
+      <bpmn:conditionExpression xsi:type="bpmn:tFormalExpression">=offeringWeeks &lt;= 17</bpmn:conditionExpression>
+    </bpmn:sequenceFlow>
+    <bpmn:sequenceFlow id="Flow_1m2o8z2" name="Offering weeks greater than 34" sourceRef="Gateway_1gcbtrb" targetRef="Event_1e56skq">
+      <bpmn:conditionExpression xsi:type="bpmn:tFormalExpression">=offeringWeeks &gt; 34</bpmn:conditionExpression>
+    </bpmn:sequenceFlow>
+    <bpmn:sequenceFlow id="Flow_1pdrx9i" name="Offering weeks greater than 17 and less than or equals 34" sourceRef="Gateway_1gcbtrb" targetRef="Event_0xpdeqq">
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn:123]: Just to share something that is not so clear in the editor, the gateway can have a default path, as shown below.
Using this we can remove the "last" condition.

image

image

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn [hunk 16]
```diff
@@ -95,84 +93,212 @@
...
           </bpmn:lane>
         </bpmn:childLaneSet>
       </bpmn:lane>
     </bpmn:laneSet>
-    <bpmn:intermediateThrowEvent id="Event_123oime" name="calculatedDataTotalEligibleDependants">
+    <bpmn:sequenceFlow id="Flow_19cfg17" sourceRef="Gateway_1blaf95" targetRef="Event_1j6ep9h" />
+    <bpmn:sequenceFlow id="Flow_0wfngj4" sourceRef="Event_1dvh1rf" targetRef="Gateway_1blaf95" />
+    <bpmn:sequenceFlow id="Flow_04xt4on" sourceRef="Event_1e56skq" targetRef="Gateway_1blaf95" />
+    <bpmn:sequenceFlow id="Flow_173rbte" sourceRef="Event_0xpdeqq" targetRef="Gateway_1blaf95" />
+    <bpmn:sequenceFlow id="Flow_0mlhznt" name="Offering weeks less than or equals 17" sourceRef="Gateway_1gcbtrb" targetRef="Event_1dvh1rf">
```
Comment (andrewsignori-aot) [inline sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn:117]: Maybe we can avoid repeating the logic in the below calculations and have one more variable created to store 17, 34, or 52.

image

**Review by andrewsignori-aot — COMMENTED** [review]
Great work and thanks for the calculation centralization for `Calculate Eligible Dependents`.
Please take a look at the comments.

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn (hunk 2 above)
Comment (andrepestana-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn]: For me, if they didn't become 12, they are 11. This "11 years until they become 12 years old" only confuses me and even though I understand what you say, I understand that if they are 11.9999 years old, they are still 11y.o.

**Comment by sh16011993** [comment]
> I have a question from the story itself:
> 
> image
> Here, the first two lines say which of the dependents are eligible. It says 12+ with disability are eligible.
> 
> And below the expectation removes disability and includes tax:
> 
> image
> Can you please clarify on this @dheepak-aot. Maybe I am missing something. Thanks!

Thanks for the clarification @dheepak-aot

**Comment by sh16011993** [comment]
> > I have a question from the story itself:
> > image
> > Here, the first two lines say which of the dependents are eligible. It says 12+ with disability are eligible.
> > And below the expectation removes disability and includes tax:
> > image
> > Can you please clarify on this @dheepak-aot. Maybe I am missing something. Thanks!
> 
> @sh16011993 as discussed, the `declaredOnTaxes` here means declared on taxes for disability and hence it does not remove the condition for disability

Thanks for the clarification @dheepak-aot

File: sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts (hunk 7 above)
Comment (sh16011993) [reply sources/packages/backend/workflow/test/2022-2023/parttime-assessment/costs/parttime-assessment-2022-2023-costs-child-care-cost.e2e-spec.ts:123]: Thanks for the clarification @dheepak-aot

File: sources/packages/backend/workflow/test/2023-2024/parttime-assessment/costs/parttime-assessment-2023-2024-costs-child-care-cost.e2e-spec.ts (hunk 9 above)
Comment (sh16011993) [reply sources/packages/backend/workflow/test/2023-2024/parttime-assessment/costs/parttime-assessment-2023-2024-costs-child-care-cost.e2e-spec.ts:123]: https://github.com/bcgov/SIMS/pull/2816#discussion_r1486887810

File: sources/packages/backend/workflow/test/2021-2022/parttime-assessment/costs/parttime-assessment-2021-2022-costs-child-care-cost.e2e-spec.ts (hunk 8 above)
Comment (sh16011993) [reply sources/packages/backend/workflow/test/2021-2022/parttime-assessment/costs/parttime-assessment-2021-2022-costs-child-care-cost.e2e-spec.ts:123]: https://github.com/bcgov/SIMS/pull/2816#discussion_r1486887810

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn (hunk 14 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn:251]: I am not able to follow how can we use MAX here as they are 2 different variables.

image

Let me know if I am missing something here.

Comment (andrewsignori-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn:251]: Never mind, my bad.

**Review by sh16011993 — APPROVED** [review]
Nice work @dheepak-aot . Great test coverage 👍

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn (hunk 2 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2021-2022.bpmn]: I am ok to change the comment as `// Dependant who is between 0 and 11 years as on study start date.` assuming it is easy to understand.

**Review by andrepestana-aot — APPROVED** [review]
Looks good! Thanks for the changes.

File: sources/packages/forms/src/form-definitions/sfaa2021-22.json [hunk 17]
```diff
@@ -26391,10 +26388,10 @@
           "key": "dependantCareCostsPanel",
           "tags": [],
           "properties": {},
-          "customConditional": "",
+          "customConditional": "show = (data.howWillYouBeAttendingTheProgram === \"Full Time\" \n            || data.hasDependents === \"yes\");",
```
Comment (lewischen-aot) [inline sources/packages/forms/src/form-definitions/sfaa2021-22.json:26391]: I think this condition will display the Dependent Care Cost panel and the first question when the intensity is full time even though the student doesn't have any dependent. We probably want to remove the "full time" condition?

Comment (dheepak-aot) [reply sources/packages/forms/src/form-definitions/sfaa2021-22.json:26391]: @lewischen-aot we need the intensity condition because, the following question which is part of the panel needs to be displayed for FT, even though the student does not have any dependent.

image

File: sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn (hunk 15 above)
Comment (dheepak-aot) [reply sources/packages/backend/workflow/BPMN/camunda-8/parttime-assessment-2023-2024.bpmn:125]: Leaving it as it is as discussed expecting gateway to throw error if no condition is satisfied.

**Review by andrewsignori-aot — APPROVED** [review]
Thanks for doing the changes, great work, and thanks for the workflow improvements. Looks good 👍

**Review by guru-aot — APPROVED** [review]
LGTM, nice work @dheepak-aot

**Review by lewischen-aot — APPROVED** [review]
LGTM. Great work @dheepak-aot
